crates/oxc_linter/src/service/runtime.rs

```diff
@@ -102,105 +191,424 @@ impl Runtime {
         })
     }
 
+    /// Prepare entry modules for linting.
+    ///
+    /// `on_entry` is called for each entry modules in `self.paths` when it's ready for linting,
+    /// which means all its dependencies are resolved if import plugin is enabled.
+    fn resolve_modules<'a>(
+        &'a mut self,
+        scope: &Scope<'a>,
+        check_syntax_errors: bool,
+        tx_error: &'a DiagnosticSender,
+        on_entry: impl Fn(&'a Self, EntryModule) + Send + Sync + Clone + 'a,
+    ) {
+        if self.resolver.is_none() {
+            self.paths.par_iter().for_each(|path| {
+                let output = self.process_path(Arc::clone(path), check_syntax_errors, tx_error);
+                let entry =
+                    EntryModule::from_processed_module(output.path, output.processed_module)
+                        .unwrap();
+                on_entry(self, entry);
+            });
+            return;
+        }
+        // The goal of code below is to construct the module graph bootstrapped by the entry modules (`self.paths`),
+        // and call `on_entry` when all dependencies of that entry is resolved. We want to call `on_entry` for each
+        // entry as soon as possible, so that the memory for source texts and semantics can be released early.
+
+        // Sorting paths to make deeper paths appear first.
+        // Consider a typical scenario:
+        //
+        // - src/index.js
+        // - src/a/foo.js
+        // - src/b/bar.js
+        // ..... (thousands of sources)
+        // - src/very/deep/path/baz.js
+        //
+        // All paths above are in `self.paths`. `src/index.js`, the entrypoint of the application, references
+        // almost all the other paths as its direct or indirect dependencies.
+        //
+        // If we construct the module graph starting from `src/index.js`, contents (sources and semantics) of
+        // all these paths must stay in memory (because they are both entries and part of `src/index.js` dependencies)
+        // until the last dependency is processed.
+        // The more efficient way is to start from "leaf" modules: their dependencies are ready earlier, thus we
+        // can run lint on them and then released their content earlier.
+        //
+        // But it's impossible to know which ones are "leaf" modules before parsing even starts. Here we assume
+        // deeper paths are more likely to be leaf modules  (src/very/deep/path/baz.js is likely to have
+        // fewer dependencies than src/index.js).
+        // This heuristic is not always true, but it works well enough for real world codebases.
+        self.paths.par_sort_unstable_by(|a, b| Path::new(b).cmp(Path::new(a)));
```

Comment:
Sort by `Path` is going to be slow. Maybe sort by number of `/` or `\\` with `sort_by_cached_key`?

Reply:
I tried to sort by counting `/` or `\\` and found the memory usage regressed a lot. I'm 100% sure why but my guess is it's important that files under same folders are near each other, because they are more likely to be in a same module graph. Processing files under same folders in a group reduces the possibility of prematurely discovering other modules to lint.

Profiling shows that the time of this sort is just a tiny fraction of the whole workflow. I think we can keep it for the better grouping strategy.